cpp/src/arrow/compute/kernels/select_k_test.cc

```diff
@@ -0,0 +1,714 @@
+enable_if_t<TypeTraits<ArrowType>::is_parameter_free, std::shared_ptr<DataType>>
+TypeToDataType() {
+  return TypeTraits<ArrowType>::type_singleton();
+}
+
+template <typename ArrowType>
+enable_if_t<std::is_same<ArrowType, TimestampType>::value, std::shared_ptr<DataType>>
+TypeToDataType() {
+  return timestamp(TimeUnit::MILLI);
+}
+
+template <typename ArrowType>
+enable_if_t<std::is_same<ArrowType, Time32Type>::value, std::shared_ptr<DataType>>
+TypeToDataType() {
+  return time32(TimeUnit::MILLI);
+}
+
+template <typename ArrowType>
+enable_if_t<std::is_same<ArrowType, Time64Type>::value, std::shared_ptr<DataType>>
+TypeToDataType() {
+  return time64(TimeUnit::NANO);
+}
+
+// ----------------------------------------------------------------------
+// Tests for SelectK
+
+template <typename ArrayType>
+auto GetLogicalValue(const ArrayType& array, uint64_t index)
+    -> decltype(array.GetView(index)) {
+  return array.GetView(index);
+}
+
+Decimal128 GetLogicalValue(const Decimal128Array& array, uint64_t index) {
+  return Decimal128(array.Value(index));
+}
+
+Decimal256 GetLogicalValue(const Decimal256Array& array, uint64_t index) {
+  return Decimal256(array.Value(index));
+}
+
+}  // namespace
+
+template <typename ArrayType, SortOrder order>
+class SelectKComparator {
+ public:
+  template <typename Type>
+  bool operator()(const Type& lval, const Type& rval) {
+    if (is_floating_type<typename ArrayType::TypeClass>::value) {
+      // NaNs ordered after non-NaNs
+      if (rval != rval) return true;
+      if (lval != lval) return false;
+    }
+    if (order == SortOrder::Ascending) {
+      return lval <= rval;
+    } else {
+      return rval <= lval;
+    }
+  }
+};
+
+template <SortOrder order>
+Result<std::shared_ptr<Array>> SelectK(const ChunkedArray& values, int64_t k) {
+  if (order == SortOrder::Descending) {
+    return TopK(values, k);
+  } else {
+    return BottomK(values, k);
+  }
+}
+
+template <SortOrder order>
+Result<std::shared_ptr<Array>> SelectK(const Array& values, int64_t k) {
+  if (order == SortOrder::Descending) {
+    return TopK(values, k);
+  } else {
+    return BottomK(values, k);
+  }
+}
+template <typename ArrowType>
+class TestSelectKBase : public TestBase {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+ protected:
+  void Validate(const ArrayType& array, int k, ArrayType& select_k, SortOrder order) {
+    ASSERT_OK_AND_ASSIGN(auto sorted_indices, SortIndices(array, order));
+    ASSERT_OK_AND_ASSIGN(Datum sorted_datum,
+                         Take(array, sorted_indices, TakeOptions::NoBoundsCheck()));
+    std::shared_ptr<Array> sorted_array_out = sorted_datum.make_array();
+
+    const ArrayType& sorted_array = *checked_pointer_cast<ArrayType>(sorted_array_out);
+
+    if (k < array.length()) {
+      for (uint64_t i = 0; i < (uint64_t)select_k.length(); ++i) {
+        const auto lval = GetLogicalValue(select_k, i);
+        const auto rval = GetLogicalValue(sorted_array, i);
```

Comment:
Instead of GetLogicalValue and such, you could slice the arrays and use AssertArraysEqual/AssertArraysApproxEqual. With the `verbose == true` parameter this will also print out the mismatched values on failure.

Reply:
That would also let you get rid of various casts and such.